File: PHP/One/Transforms/ResolveUnresolvedTypes.php
<?php

namespace One\Transforms\ResolveUnresolvedTypes;

use One\AstTransformer\AstTransformer;
use One\Ast\AstTypes\UnresolvedType;
use One\Ast\AstTypes\ClassType;
use One\Ast\AstTypes\InterfaceType;
use One\Ast\AstTypes\EnumType;
use One\Ast\AstTypes\GenericsType;
use One\Ast\Types\Class_;
use One\Ast\Types\Interface_;
use One\Ast\Types\Enum;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\UnresolvedNewExpression;
use One\Ast\Expressions\NewExpression;
use One\Ast\Interfaces\IType;

class ResolveUnresolvedTypes extends AstTransformer {
    function __construct() {
        parent::__construct("ResolveUnresolvedTypes");
        
    }
    
    protected function visitType($type) {
        parent::visitType($type);
        if ($type instanceof UnresolvedType) {
            if ($this->currentInterface !== null && in_array($type->typeName, $this->currentInterface->typeArguments))
                return new GenericsType($type->typeName);
            
            $symbol = $this->currentFile->availableSymbols->get($type->typeName);
            if ($symbol === null) {
                $this->errorMan->throw("Unresolved type '" . $type->typeName . "' was not found in available symbols");
                return $type;
            }
            
            if ($symbol instanceof Class_)
                return new ClassType($symbol, $type->typeArguments);
            else if ($symbol instanceof Interface_)
                return new InterfaceType($symbol, $type->typeArguments);
            else if ($symbol instanceof Enum)
                return new EnumType($symbol);
            else {
                $this->errorMan->throw("Unknown symbol type: " . $symbol);
                return $type;
            }
        }
        else
            return $type;
    }
    
    protected function visitExpression($expr) {
        if ($expr instanceof UnresolvedNewExpression) {
            $clsType = $this->visitType($expr->cls);
            if ($clsType instanceof ClassType) {
                $newExpr = new NewExpression($clsType, $expr->args);
                $newExpr->parentNode = $expr->parentNode;
                parent::visitExpression($newExpr);
                return $newExpr;
            }
            else {
                $this->errorMan->throw("Excepted ClassType, but got " . $clsType);
                return $expr;
            }
        }
        else
            return parent::visitExpression($expr);
    }
}

File: CSharp/One/Transforms/ResolveUnresolvedTypes.cs
using One.Ast;
using One;

namespace One.Transforms
{
    public class ResolveUnresolvedTypes : AstTransformer
    {
        public ResolveUnresolvedTypes(): base("ResolveUnresolvedTypes")
        {
            
        }
        
        protected override IType visitType(IType type)
        {
            base.visitType(type);
            if (type is UnresolvedType unrType) {
                if (this.currentInterface != null && this.currentInterface.typeArguments.includes(unrType.typeName))
                    return new GenericsType(unrType.typeName);
                
                var symbol = this.currentFile.availableSymbols.get(unrType.typeName);
                if (symbol == null) {
                    this.errorMan.throw_($"Unresolved type '{unrType.typeName}' was not found in available symbols");
                    return unrType;
                }
                
                if (symbol is Class class_)
                    return new ClassType(class_, unrType.typeArguments);
                else if (symbol is Interface int_)
                    return new InterfaceType(int_, unrType.typeArguments);
                else if (symbol is Enum_ enum_)
                    return new EnumType(enum_);
                else {
                    this.errorMan.throw_($"Unknown symbol type: {symbol}");
                    return unrType;
                }
            }
            else
                return type;
        }
        
        protected override Expression visitExpression(Expression expr)
        {
            if (expr is UnresolvedNewExpression unrNewExpr) {
                var clsType = this.visitType(unrNewExpr.cls);
                if (clsType is ClassType classType) {
                    var newExpr = new NewExpression(classType, unrNewExpr.args);
                    newExpr.parentNode = unrNewExpr.parentNode;
                    base.visitExpression(newExpr);
                    return newExpr;
                }
                else {
                    this.errorMan.throw_($"Excepted ClassType, but got {clsType}");
                    return unrNewExpr;
                }
            }
            else
                return base.visitExpression(expr);
        }
    }
}